src/Microsoft.IdentityModel.Aad/ClaimsTypeAccessor.cs

```diff
@@ -0,0 +1,258 @@
+using System.Security.Claims;
+using System.Linq;
+
+namespace Microsoft.IdentityModel.Aad
+{
+    /// <summary>
+    /// Defines extension methods to access claims by purpose, and not by name, making it compatible
+    /// with several versions of tokens.
+    /// </summary>
+    public static partial class ClaimPrincipalExtensions
+    {
+        /// <summary> 
+        /// Verifies that the token has any of the claims required. 
+        /// </summary> 
+        /// <param name="user">claims identity</param> 
+        /// <param name="requiredClaims">Type (name) of the required claims</param> 
+        /// <returns></returns> 
+        public static bool ContainsAnyClaim(this ClaimsPrincipal user, params string[] requiredClaims)
+        {
+            if (user == null)
+            {
+                throw new ArgumentNullException(nameof(user));
+            }
+            return user.Claims.Any(claim => requiredClaims.Contains(claim.Type));
+        }
+
+        /// <summary>
+        /// Get the first value of a claim or its equivalent
+        /// </summary>
+        /// <param name="user">Claims principal</param>
+        /// <param name="claimType">Claims type</param>
+        /// <returns></returns>
+        public static string GetEquivalentClaim(this ClaimsPrincipal user, string claimType)
+        {
+            if (equivalentClaims.TryGetValue(claimType, out string[] equivalents))
+            {
+                return GetClaimValue(user, equivalents);
+            }
+            else
+            {
+                return null;
+            }    
+        }
+
+        /// <summary>
+        /// Get the first value of a claim or its equivalent
+        /// </summary>
+        /// <param name="claimsIdentity">Claims identity.</param>
+        /// <param name="claimType">Claims type</param>
+        /// <returns></returns>
+        public static string GetEquivalentClaim(this ClaimsIdentity claimsIdentity, string claimType)
+        {
+            if (equivalentClaims.TryGetValue(claimType, out string[] equivalents))
+            {
+                return GetClaimValue(claimsIdentity, equivalents);
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Get the first value of a claim or its equivalent
+        /// </summary>
+        /// <param name="user">Claims principal</param>
+        /// <param name="claimType">Claims type.</param>
+        /// <returns>All the values of the claims or its equivalents.</returns>
+        public static IEnumerable<string> GetAllEquivalentClaims(this ClaimsPrincipal user, string claimType)
+        {
+            if (user is null)
+            {
+                throw new ArgumentNullException(nameof(user));
+            }
+
+            if (string.IsNullOrEmpty(claimType))
+            {
+                throw new ArgumentException($"'{nameof(claimType)}' cannot be null or empty.", nameof(claimType));
```

Comment:
We have been using a single location for log messages, each with an identifier that allows for quickly identifying the assembly and location of the issue.
see: [LogMessages](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/050297acc5b1dacba8f706459af6da8803762453/src/Microsoft.IdentityModel.Tokens/LogMessages.cs#L36)